File: server/models/destination.model.js
import {DataTypes} from 'sequelize';
import sequelize from '../database/db.js';

const Destination = sequelize.define('Destination', {
    id: {
        type: DataTypes.UUID,
        defaultValue: DataTypes.UUIDV4,
        primaryKey: true,
    },
    categoryId: {
        type: DataTypes.UUID,
        allowNull: false,
        references: {
            model: 'destination_categories',
            key: 'id'
        }
    },
    name: {
        type: DataTypes.STRING(255),
        allowNull: false,
        index: true // Thêm index để tìm kiếm nhanh
    },
    description: {
        type: DataTypes.TEXT,
        allowNull: true,
    },
    address: {
        type: DataTypes.STRING(255),
        allowNull: true,
    },
    // Sử dụng PostGIS cho query không gian nâng cao
    lat: {
        type: DataTypes.FLOAT,
        allowNull: false,
    },
    lng: {
        type: DataTypes.FLOAT,
        allowNull: false,
    },
    // Thêm trường geometry cho PostGIS
    geom: {
        type: DataTypes.GEOMETRY('POINT'),
        allowNull: true,
    },
    avg_cost: {
        type: DataTypes.DECIMAL(12, 2),
        defaultValue: 0,
    },
    rating: {
        type: DataTypes.FLOAT,
        defaultValue: 0,
        validate: {min: 0, max: 5}
    },
    total_reviews: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    total_likes: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    total_checkins: {
        type: DataTypes.INTEGER,
        defaultValue: 0,
    },
    tags: {
        type: DataTypes.ARRAY(DataTypes.TEXT),
        defaultValue: []
    },
    opening_hours: {
        type: DataTypes.JSONB,
        allowNull: true,
    },
    images: {
        type: DataTypes.ARRAY(DataTypes.STRING),
        defaultValue: []
    },
    // Thêm các trường mới cho AI và gamification
    ai_summary: {
        type: DataTypes.TEXT,
        allowNull: true,
        comment: 'Tóm tắt AI về địa điểm'
    },
    best_time_to_visit: {
        type: DataTypes.STRING(100),
        allowNull: true,
        comment: 'Thời gian tốt nhất để ghé thăm'
    },
    recommended_duration: {
        type: DataTypes.INTEGER,
        defaultValue: 60,
        comment: 'Thời gian đề xuất (phút)'
    },
    is_hidden_gem: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
        comment: 'Có phải là địa điểm ít người biết không'
    },
    challenge_tags: {
        type: DataTypes.ARRAY(DataTypes.STRING),
        defaultValue: [],
        comment: 'Tag cho thử thách gamification'
    },
    seasonality: {
        type: DataTypes.JSONB,
        allowNull: true,
        comment: 'Thông tin mùa vụ: {summer: true, winter: false, ...}'
    },
    accessibility: {
        type: DataTypes.JSONB,
        allowNull: true,
        comment: 'Thông tin tiếp cận: {wheelchair: true, parking: true, ...}'
    },
    contact_info: {
        type: DataTypes.JSONB,
        allowNull: true,
        comment: 'Thông tin liên hệ: {phone: "...", website: "...", email: "..."}'
    },
    social_media: {
        type: DataTypes.JSONB,
        allowNull: true,
        comment: 'Mạng xã hội: {facebook: "...", instagram: "...", tiktok: "..."}'
    },
    is_verified: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
        comment: 'Đã được xác minh'
    },
    is_featured: {
        type: DataTypes.BOOLEAN,
        defaultValue: false,
        comment: 'Địa điểm nổi bật'
    },
    is_active: {
        type: DataTypes.BOOLEAN,
        defaultValue: true
    },
    metadata: {
        type: DataTypes.JSONB,
        defaultValue: {},
        comment: 'Metadata cho AI và analytics'
    }
}, {
    tableName: 'destinations',
    timestamps: true,
    // Thêm hooks để tự động tạo geometry từ lat/lng
    hooks: {
        beforeCreate: async (destination, options) => {
            if (destination.lat && destination.lng) {
                destination.geom = sequelize.fn('ST_SetSRID',
                    sequelize.fn('ST_MakePoint', destination.lng, destination.lat),
                    4326
                );
            }
        },
        beforeUpdate: async (destination, options) => {
            if (destination.changed('lat') || destination.changed('lng')) {
                destination.geom = sequelize.fn('ST_SetSRID',
                    sequelize.fn('ST_MakePoint', destination.lng, destination.lat),
                    4326
                );
            }
        }
    }
});

export default Destination;

File: client/lib/features/destinations/data/models/destination_model.dart
import '../../domain/entities/destination.dart';

class DestinationModel extends Destination {
  const DestinationModel({
    required super.id,
    required super.categoryId,
    required super.name,
    required super.description,
    required super.address,
    required super.lat,
    required super.lng,
    super.avgCost,
    required super.rating,
    required super.totalReviews,
    required super.totalLikes,
    required super.totalSaves,
    required super.totalCheckins,
    required super.tags,
    super.openingHours,
    required super.images,
    super.aiSummary,
    super.bestTimeToVisit,
    super.recommendedDuration,
    required super.isHiddenGem,
    required super.challengeTags,
    required super.isVerified,
    required super.isFeatured,
    required super.isActive,
    required super.createdAt,
    required super.updatedAt,
    super.category,
    super.isLiked,
  });

  factory DestinationModel.fromJson(Map<String, dynamic> json) {
    return DestinationModel(
      id: json['id'] ?? '',
      categoryId: json['categoryId'] ?? '',
      name: json['name'] ?? '',
      description: json['description'] ?? '',
      address: json['address'] ?? '',
      lat: (json['lat'] as num?)?.toDouble() ?? 0.0,
      lng: (json['lng'] as num?)?.toDouble() ?? 0.0,
      avgCost: json['avg_cost'] != null ? double.tryParse(json['avg_cost'].toString()) : null,
      rating: (json['rating'] as num?)?.toDouble() ?? 0.0,
      totalReviews: json['total_reviews'] ?? 0,
      totalLikes: json['total_likes'] ?? 0,
      totalSaves: json['total_saves'] ?? 0,
      totalCheckins: json['total_checkins'] ?? 0,
      tags: (json['tags'] as List?)?.map((e) => e.toString()).toList() ?? [],
      openingHours:
          json['opening_hours'] != null ? Map<String, String>.from(json['opening_hours']) : null,
      images: (json['images'] as List?)?.map((e) => e.toString()).toList() ?? [],
      aiSummary: json['ai_summary'],
      bestTimeToVisit: json['best_time_to_visit'],
      recommendedDuration: json['recommended_duration'],
      isHiddenGem: json['is_hidden_gem'] ?? false,
      challengeTags: (json['challenge_tags'] as List?)?.map((e) => e.toString()).toList() ?? [],
      isVerified: json['is_verified'] ?? false,
      isFeatured: json['is_featured'] ?? false,
      isActive: json['is_active'] ?? true,
      createdAt: DateTime.parse(json['createdAt'] ?? DateTime.now().toIso8601String()),
      updatedAt: DateTime.parse(json['updatedAt'] ?? DateTime.now().toIso8601String()),
      category:
          json['category'] != null ? DestinationCategoryModel.fromJson(json['category']) : null,
      isLiked: json['isLiked'] as bool?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'categoryId': categoryId,
      'name': name,
      'description': description,
      'address': address,
      'lat': lat,
      'lng': lng,
      'avg_cost': avgCost,
      'rating': rating,
      'total_reviews': totalReviews,
      'total_likes': totalLikes,
      'total_saves': totalSaves,
      'total_checkins': totalCheckins,
      'tags': tags,
      'opening_hours': openingHours,
      'images': images,
      'ai_summary': aiSummary,
      'best_time_to_visit': bestTimeToVisit,
      'recommended_duration': recommendedDuration,
      'is_hidden_gem': isHiddenGem,
      'challenge_tags': challengeTags,
      'is_verified': isVerified,
      'is_featured': isFeatured,
      'is_active': isActive,
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'category': category != null
          ? DestinationCategoryModel(
              id: category!.id,
              name: category!.name,
              icon: category!.icon,
              description: category!.description,
              travelStyleId: category!.travelStyleId,
              contextTags: category!.contextTags,
              popularityScore: category!.popularityScore,
              avgVisitDuration: category!.avgVisitDuration,
              isActive: category!.isActive,
              createdAt: category!.createdAt,
              updatedAt: category!.updatedAt,
            ).toJson()
          : null,
    };
  }
}

class DestinationCategoryModel extends DestinationCategory {
  const DestinationCategoryModel({
    required super.id,
    required super.name,
    required super.icon,
    super.description,
    required super.travelStyleId,
    required super.contextTags,
    super.popularityScore,
    super.avgVisitDuration,
    required super.isActive,
    super.createdAt,
    super.updatedAt,
  });

  factory DestinationCategoryModel.fromJson(Map<String, dynamic> json) {
    return DestinationCategoryModel(
      id: json['id'] ?? '',
      name: json['name'] ?? '',
      icon: json['icon'] ?? '',
      description: json['description'],
      travelStyleId: json['travel_style_id'] ?? '',
      contextTags: (json['context_tags'] as List?)?.map((e) => e.toString()).toList() ?? [],
      popularityScore: (json['popularity_score'] as num?)?.toDouble(),
      avgVisitDuration: json['avg_visit_duration'],
      isActive: json['is_active'] ?? true,
      createdAt: json['createdAt'] != null ? DateTime.parse(json['createdAt']) : null,
      updatedAt: json['updatedAt'] != null ? DateTime.parse(json['updatedAt']) : null,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'icon': icon,
      'description': description,
      'travel_style_id': travelStyleId,
      'context_tags': contextTags,
      'popularity_score': popularityScore,
      'avg_visit_duration': avgVisitDuration,
      'is_active': isActive,
      'createdAt': createdAt?.toIso8601String(),
      'updatedAt': updatedAt?.toIso8601String(),
    };
  }
}
